coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java

```diff
@@ -37,80 +45,123 @@
-            QueryImplementor queryObj = query.unwrap(QueryImplementor.class);
-            return queryObj.getQueryString();
-        } catch (Exception e) {
-            throw new TechnicalException(CoffeeFaultType.OPERATION_FAILED, "Failed to unwrap QueryImpl from Query", e);
-        }
+    /**
+     * Get native SQL from created Query
+     *
+     * @param em
+     *            entity manager (ignored)
+     * @param criteriaQuery
+     *            created query
+     * @return native SQL of query
+     * @throws BaseException
+     *             on error
+     * @deprecated use {@link #toNativeSQL(Query)} instead
+     */
+    @Deprecated(since = "2.1.0", forRemoval = true)
+    public static String toNativeSQL(EntityManager em, Query criteriaQuery) throws BaseException {
+        return toNativeSQL(criteriaQuery);
     }
 
-    private static String getSQLString(EntityManager em, String jpqlString) throws BaseException {
-        try {
-            Object sessionImpl = em.unwrap(Class.forName("org.hibernate.internal.SessionImpl"));
-            Class<?> clazzInner = sessionImpl.getClass().getSuperclass();
-            Class<?> clazz = clazzInner.getSuperclass();
-            Method m = clazz.getDeclaredMethod("getQueryPlan", new Class[] { String.class, boolean.class });
-            m.setAccessible(true);
-            Object hqlQueryPlan = m.invoke(sessionImpl, new Object[] { jpqlString, false });
-
-            Class<?> planClazz = hqlQueryPlan.getClass();
-            Method plm = planClazz.getMethod("getSqlStrings");
-            String[] res = (String[]) plm.invoke(hqlQueryPlan);
-            return res[0];
-        } catch (Exception e) {
-            throw new TechnicalException(CoffeeFaultType.OPERATION_FAILED, "Getting native SQL from JPQL failed", e);
+    /**
+     * Get native SQL from created Query
+     *
+     * @param criteriaQuery
+     *            created query
+     * @return native SQL of query
+     * @throws BaseException
+     *             on error
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static String toNativeSQL(Query criteriaQuery) throws BaseException {
+        QueryImplementor query = criteriaQuery.unwrap(QueryImplementor.class);
+        if (query instanceof SqmInterpretationsKey.InterpretationsKeySource && query instanceof QuerySqmImpl) {
+            QueryInterpretationCache.Key cacheKey = SqmInterpretationsKey
+                    .createInterpretationsKey((SqmInterpretationsKey.InterpretationsKeySource) query);
+            QuerySqmImpl<?> querySqm = (QuerySqmImpl<?>) query;
+            Supplier buildSelectQueryPlan = () -> {
+                try {
+                    return ReflectionUtils.invokeMethod(querySqm, "buildSelectQueryPlan", ConcreteSqmSelectQueryPlan.class);
+                } catch (BaseException e) {
+                    Logger.getLogger(JpaUtil.class).warn("Exception on calling buildSelectQueryPlan()!", e);
+                    return null;
```

Comment:
why doesn't it throw the exception but return null instead?

Reply:
Supplier<T> cannot throw checked exception and next step needs the Supplier<T> interface as parameter (so i couldnt use own supplier with baseexception )